### ISSUE TITLE
Unable to build on Windows

### ISSUE THREAD

**Issue by enko** [holepunchto/sodium-native#79]
I tried to install sodium-native as a dependency and it failed because the process can not find `msbuild.exe`. I then remembered that I need to install https://github.com/felixrieseberg/windows-build-tools which I did, but still no avail. After a bit digging I found out that on my computer msbuild.exe was now under `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\msbuild.exe`. I applied this patch to preinstall.js and the build process worked fine:

```patch
diff --git a/preinstall.js b/preinstall.js
index 733d483..8a39db6 100644
--- a/preinstall.js
+++ b/preinstall.js
@@ -65,7 +65,7 @@ function buildWindows () {
 
   spawn('.\\msvc-scripts\\process.bat', [], {cwd: dir, stdio: 'inherit'}, function (err) {
     if (err) throw err
-    var msbuild = path.resolve('/', 'Program Files (x86)', 'MSBuild/14.0/Bin/MSBuild.exe')
+    var msbuild = path.resolve('/', 'Program Files (x86)', 'Microsoft Visual Studio', '2017/BuildTools/MSBuild/15.0/Bin/MSBuild.exe')
     var args = ['/p:Configuration=ReleaseDLL;Platform=' + warch, '/nologo']
     spawn(msbuild, args, {cwd: dir, stdio: 'inherit'}, function (err) {
       if (err) throw err
```

I am not an expert in this matter, but I think this is because the default of `windows-build-tools` was changed from VS2015 to VS2017:

https://github.com/felixrieseberg/windows-build-tools/commit/698d86c944ce3c86e02dd73e41f808147531db02

If you need any information please let me know.

**Comment by enko** [comment]
Looking a bit further I could port this method into the preinstall script:

```powershell
Function Find-MsBuild([int] $MaxVersion = 2017)
{
    $agentPath = "$Env:programfiles (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\msbuild.exe"
    $devPath = "$Env:programfiles (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\msbuild.exe"
    $proPath = "$Env:programfiles (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\msbuild.exe"
    $communityPath = "$Env:programfiles (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\msbuild.exe"
    $fallback2015Path = "${Env:ProgramFiles(x86)}\MSBuild\14.0\Bin\MSBuild.exe"
    $fallback2013Path = "${Env:ProgramFiles(x86)}\MSBuild\12.0\Bin\MSBuild.exe"
    $fallbackPath = "C:\Windows\Microsoft.NET\Framework\v4.0.30319"
		
    If ((2017 -le $MaxVersion) -And (Test-Path $agentPath)) { return $agentPath } 
    If ((2017 -le $MaxVersion) -And (Test-Path $devPath)) { return $devPath } 
    If ((2017 -le $MaxVersion) -And (Test-Path $proPath)) { return $proPath } 
    If ((2017 -le $MaxVersion) -And (Test-Path $communityPath)) { return $communityPath } 
    If ((2015 -le $MaxVersion) -And (Test-Path $fallback2015Path)) { return $fallback2015Path } 
    If ((2013 -le $MaxVersion) -And (Test-Path $fallback2013Path)) { return $fallback2013Path } 
    If (Test-Path $fallbackPath) { return $fallbackPath } 
        
    throw "Yikes - Unable to find msbuild"
}
```

https://alastaircrabtree.com/how-to-find-latest-version-of-msbuild-in-powershell/

whats your opinion on that?

**Comment by emilbayes** [comment]
We were actually adding a "findMsbuild" function to the build script as our Windows build doesn't work on the new travis windows. Feel free to open a PR if you want credit :)

Do you by any chance know how to handle this case as well? https://github.com/sodium-friends/sodium-native/pull/74

Is that handled by the `$Env:programfiles`?

**Comment by enko** [comment]
Will submit 👍 

This will also solve #74 since the environment variable ProgramFiles also contains the drive path.

**Comment by emilbayes** [comment]
Closed by #81 (apparently also fixes #74)
